mono/metadata/threads.c

```diff
@@ -4059,12 +4085,23 @@ mono_threads_abort_appdomain_threads (MonoDomain *domain, int timeout)
 	return TRUE;
 }
 
+/* This is a JIT icall.  This icall is called from a finally block when
+ * mono_install_handler_block_guard called by another thread has flipped the
+ * finally block's exvar (see mono_find_exvar_for_offset).  In that case, if
+ * the finally is in an abort protected block, we must defer the abort
+ * exception until we leave the abort protected block.  Otherwise we proceed
+ * with a synchronous self-abort.
+ */
 void
-mono_thread_self_abort (void)
+ves_icall_thread_finish_async_abort (void)
 {
-	ERROR_DECL (error);
-	self_abort_internal (error);
-	mono_error_set_pending_exception (error);
+	/* If we're in an abort protected block, just set the async requested bit and return.
+	 * Otherwise self abort now.
+	 */
+	if (!mono_thread_set_self_interruption_respect_abort_prot ())
+		return;
```

Comment:
Can't his be
```
if (mono_thread_set_self_interruption_respect_abort_prot ())
   mono_thread_info_self_interrupt ();
```
?

Reply:
@lambdageek why do we need to call mono_thread_info_self_interrupt here ?

Reply:
Good question. I'm not sure.

I'm copying from `self_abort_internal`:
1.  I don't need to do `mono_thread_request_interruption (TRUE)` anymore - because I already called mono_thread_set_self_interruption_respect_abort_prot` and it returned TRUE,
2. so... (looking at the body of `mono_thread_request_interruption`) the next thing I need to do is `mono_thread_info_self_interrupt()`.

Reply:
@lambdageek the comment in `mono_thread_request_interruption` says that the method can be called from signal handler. Maybe it needs to break some waits if it is called while we are in unmanaged or something like that. In this case we know we call it directly from managed code after returning from a finally clause, so I don't see any waits that would need to be broken.